projects/bootloader/src/config.c

```diff
@@ -0,0 +1,87 @@
+#include "persist.h"
+#include "soft_timer.h"
+
+static PersistStorage s_config_1_persist = { 0 };
+static BootloaderConfig s_config_1_blob = { 0 };
+
+static PersistStorage s_config_2_persist = { 0 };
+static BootloaderConfig s_config_2_blob = { 0 };
+
+#define BOOTLOADER_CONFIG_PAGE_1_FLASH_PAGE (FLASH_ADDR_TO_PAGE(BOOTLOADER_CONFIG_PAGE_1_START))
+#define BOOTLOADER_CONFIG_PAGE_2_FLASH_PAGE (FLASH_ADDR_TO_PAGE(BOOTLOADER_CONFIG_PAGE_2_START))
+
+StatusCode config_init(void) {
+  status_ok_or_return(persist_init(&s_config_1_persist, BOOTLOADER_CONFIG_PAGE_1_FLASH_PAGE,
+                                   &s_config_1_blob, sizeof(s_config_1_blob), false));
+  status_ok_or_return(persist_ctrl_periodic(&s_config_1_persist, false));
+
+  status_ok_or_return(persist_init(&s_config_2_persist, BOOTLOADER_CONFIG_PAGE_2_FLASH_PAGE,
+                                   &s_config_2_blob, sizeof(s_config_2_blob), false));
+  status_ok_or_return(persist_ctrl_periodic(&s_config_2_persist, false));
+
+  uint32_t config_1_crc = s_config_1_blob.CRC32;
+  s_config_1_blob.CRC32 = 0;
+
+  uint32_t config_2_crc = s_config_2_blob.CRC32;
+  s_config_2_blob.CRC32 = 0;
+
+  uint32_t config_1_check_crc = crc32_arr((uint8_t *)&s_config_1_blob, sizeof(s_config_1_blob));
+  uint32_t config_2_check_crc = crc32_arr((uint8_t *)&s_config_2_blob, sizeof(s_config_2_blob));
+
+  bool is_config_1_equal = config_1_crc == config_1_check_crc;
+  bool is_config_2_equal = config_2_crc == config_2_check_crc;
+
+  if (is_config_1_equal && is_config_2_equal) {
+    s_config_1_blob.CRC32 = config_1_crc;
+    s_config_2_blob.CRC32 = config_2_crc;
+  } else if (!is_config_1_equal && !is_config_2_equal) {
+    LOG_CRITICAL("ERROR: Both config pages corrupted \n");
+    return STATUS_CODE_INTERNAL_ERROR;
+  } else if (is_config_1_equal) {
+    s_config_1_blob.CRC32 = config_1_crc;
+    memcpy(&s_config_2_blob, &s_config_1_blob, sizeof(s_config_1_blob));
+    persist_commit(&s_config_2_persist);
+  } else {
+    s_config_2_blob.CRC32 = config_2_crc;
+    memcpy(&s_config_1_blob, &s_config_2_blob, sizeof(s_config_2_blob));
+    persist_commit(&s_config_1_persist);
+  }
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode config_get(BootloaderConfig *config) {
+  // Please persist any changes
+  memcpy(config, &s_config_1_blob, sizeof(s_config_1_blob));
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode config_commit(BootloaderConfig *input_config) {
+  input_config->CRC32 = 0;
+  uint32_t input_config_crc = crc32_arr((uint8_t *)input_config, sizeof(input_config));
+  input_config->CRC32 = input_config_crc;
+
+  memcpy(&s_config_1_blob, input_config, sizeof(*input_config));
+  persist_commit(&s_config_1_persist);
+
+  status_ok_or_return(persist_init(&s_config_1_persist, BOOTLOADER_CONFIG_PAGE_1_FLASH_PAGE,
+                                   &s_config_1_blob, sizeof(s_config_1_blob), false));
+
+  bool input_equals_config_1 = memcmp(&s_config_1_blob, input_config, sizeof(*input_config)) == 0;
```

Comment:
This is also kind of a cryptic name eh? IMO something like `config_1_written_correctly` would be better, or even invert the polarity and call it `config_1_corrupted`.